### ISSUE TITLE
Rate limits

### ISSUE THREAD

**Issue by iomintz** [iomintz/acnh-api#2]
We want to avoid spamming the API or otherwise making "unnatural" requests. Time how long it takes for each function to run on a real Switch and make a similar rate limit.

**Comment by badjoerichards** [comment]
Average 3.7s between each manual Switch call to dodo-code endpoint.

Made 10 calls on my patched Switch manually.
When a dodo-code fails, the fastest time it takes for me to attempt to repeat connect to the same dodo-code, by hitting A quickly through the menu to re-connect, is averaged around 3.7s.

**Comment by iomintz** [comment]
Thanks! A global ratelimit of 3.7s for all users of the API would probably be unacceptable though. I wonder if I should have one switch account configured per concurrent connection that the server can handle. That way it could distribute the requests across all configured accounts. That sounds expensive though as each of them needs a Nintendo Switch online membership.

**Comment by badjoerichards** [comment]
Yeah perhaps a combination of these things:
1. Ability for Administrator to configure multiple accounts for ACNH-API to run with the 3.7s per call ratelimit (round-robin)
2. The ACNH-API that YOU run is SPECIAL and UNIQUE (let's call it MOTHER) - and there's only 1 MOTHER in the world.
3. Other Administrators who host their own ACNH-API, can/must contribute, by processing calls for MOTHER, when MOTHER runs out of accounts to process requests, due to the 3.7s ratelimit.

That way, you don't have to deal with the costs, and the more people who use ACNH-API, will power the MOTHER ACNH-API further.

just some ideas

**Comment by iomintz** [comment]
That's a very interesting idea. I wouldn't want to centralize it so much though. Probably the other instances should all talk to each other rather than to some big bad central authority. Also keep in mind, the instances would have to trust the responses from the other ones. Someone could set up a bad apple that censors certain islands or certain designs (the latter I'm more concerned about).

**Comment by badjoerichards** [comment]
ah.. yeah! decentralized, peer-2-peer model perhaps.

EVERY ACNH-API host instance, will then help to power the whole "cloud" network further.

Indeed the trust issue is a problem to fix!
Maybe similar to Bitcoin/Crypto, where a NON-LOCAL REQUEST is processed by random REMOTE INSTANCES (2 or 3?), to confirm RESULTS and also find bad REMOTE apples. (local and central blacklist)

Phase 1: 
ACNH-API can start with ability for Administrator to configure multiple accounts to run with 3.7s per call ratelimit (round-robin)

Then Phase 2:
Consider expanding into a P2P network model - to lower costs for everyone and keep the service up worldwide.

**Comment by iomintz** [comment]
That approach is better but now requires 3× the resources. Also it's still an open question whether the same ticket can be reused on multiple accounts simultaneously without getting banned. It's possible that one ticket would be needed *per* independent request, not per as many accounts as can fit on one switch. In which case you'd need to buy one switch and one copy of AC:NH per independent request. Which is totally infeasible unless API access is sold.

**Comment by badjoerichards** [comment]
Ah yeah, you're right. If we're trying to prevent bans and ratelimits, it's proper true to make sure: 
1 "configured account" = 1 rooted Switch + 1 ACNH game + 1 NSO account

But yes, I agree this is not really feasible.

**Comment by iomintz** [comment]
Now it's *also* entirely possible that they *don't* ban for flooding the API. Consider that when you open the "posted designs" page in the Custom Designs kiosk, the game makes one full HTTP request per design you've posted, which means up to 120 rapid-fire requests. Maybe they rate limit the Dodo Code API but not the Designs API. (Also, Dodo Codes use the matchmaking servers, which operate on a completely different transport level protocol, PRUDP, so maybe it's a whole different ball game?). I think for now I'll implement no rate limits. If I get banned then I'll buy a new Switch and try again with rate limits. If I'm still not banned in a few months then it's probably safe to assume they won't ban me.

**Comment by badjoerichards** [comment]
Make sense, maybe the api ban has not been implemented yet.

If that is so, does it mean we can run multiple acnh-api webservers/instances , using the same “configured account”?

**Comment by iomintz** [comment]
Yes it would mean that. Do you have Discord? I can help you set up your mitm proxy.

**Comment by badjoerichards** [comment]
> Yes it would mean that. Do you have Discord? I can help you set up your mitm proxy.

Yes: louvre2#6166
